nominal/nominal.py

```diff
@@ -1,628 +1,316 @@
+    timestamp_type: TimestampColumnType,
+    *,
+    wait_until_complete: bool = True,
+) -> Dataset:
+    """Create a dataset in the Nominal platform from a polars.DataFrame."""
+    conn = get_default_connection()
+
+    def write_and_close(df: pl.DataFrame, w: BinaryIO) -> None:
+        df.write_csv(w)
+        w.close()
+
+    with reader_writer() as (reader, writer):
+        # write the dataframe to CSV in another thread
+        t = Thread(target=write_and_close, args=(df, writer))
+        t.start()
+        dataset = conn.create_dataset_from_io(
+            reader,
+            name,
+            timestamp_column=timestamp_column,
+            timestamp_type=timestamp_type,
+            file_type=FileTypes.CSV,
+            description=description,
+        )
+        t.join()
+    if wait_until_complete:
+        dataset.poll_until_ingestion_completed()
+    return dataset
+
+
+def upload_dataset(
+    path: Path | str,
+    name: str,
+    description: str,
+    timestamp_column: str,
+    timestamp_type: TimestampColumnType,
+    *,
+    wait_until_complete: bool = True,
+) -> Dataset:
+    """Create a dataset in the Nominal platform from a .csv, .csv.gz, or .parquet file."""
+    path = Path(path)
+    file_type = FileType.from_path_dataset(path)
+    conn = get_default_connection()
+    with open(path, "rb") as f:
+        dataset = conn.create_dataset_from_io(
+            f,
+            name,
+            timestamp_column=timestamp_column,
+            timestamp_type=timestamp_type,
+            file_type=file_type,
+            description=description,
         )
+    if wait_until_complete:
+        dataset.poll_until_ingestion_completed()
+    return dataset
 
-        if resp.status_code == 200:
-            self.s3_path = resp.text.strip('"')
-            print("\nUpload to S3 successful. S3 bucket:", self.s3_path)
-        elif resp.status_code == 401:
-            _auth_help_blurb()
-        else:
-            print("\n{0} error during upload to S3:\n".format(resp.status_code), resp.json())
 
-        return resp
+def get_dataset_by_rid(rid: str) -> Dataset:
+    """Retrieve a dataset from the Nominal platform by its RID."""
+    conn = get_default_connection()
+    return conn.get_dataset(rid)
 
-    def upload(self, overwrite: bool = False):
-        """
-        Registers Dataset in Nominal on Nominal platform.
 
-        Endpoint:
-        /ingest/v1/trigger-ingest-v2
+def update_dataset(
+    dataset: Dataset,
+    *,
+    name: str | None = None,
+    description: str | None = None,
+    properties: Mapping[str, str] | None = None,
+    labels: Sequence[str] | None = None,
+) -> Dataset:
+    """Update dataset metadata.
+    Updates the current instance and returns it.
+
+    Note: This replaces the metadata (rather than adding to it).
+    """
+    return dataset.update(name=name, description=description, properties=properties, labels=labels)
+
+
+def create_run(
+    title: str,
+    description: str,
```

Comment:
run descriptions are not optional (as per the APIs) but we could similarly default to empty string.

I feel like run titles should always be mandatory/not autogenerated

Reply:
agreed re: run titles

I think we should make run description optional. as of right now, there are ~32k runs and 20k of them have an empty string as their description. my hottake is that the backend api should reflect this & make the field optional too 👀

Reply:
made optional